### PR TITLE
ADD HPO

### PR DESCRIPTION
**Pull request by kimdanni**
[Modify] Update optuna_hpo.py to include HPOResult, HPOConfig, search_space, and run_hpo method

This PR updates optuna_hpo.py to include the following changes:

Add HPOResult dataclass to store the results of HPO
Add HPOConfig dataclass to store the configuration of HPO
Add search_space parameter to the OptunaHPO constructor to specify the search space for HPO
Update run_hpo method to return an HPOResult object
Update run_hpo method to use HPOConfig to store the configuration of HPO
These changes improve the functionality and maintainability of the OptunaHPO class by introducing dataclasses to store the results and configuration of HPO, and by allowing users to specify the search space for HPO. The updated run_hpo method also returns an HPOResult object, which makes it easier for users to access the results of HPO.

### CODE REVIEW

File: tests/test_hpo.py [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+    return x**2 + (y + z + a + q + w + e) / 2 + (r + t + y + u + i + o) * 2
+
+
+@pytest.mark.parametrize(
+    "n_trials,  search_space, direction, sampler, pruner",
+    [
+        (
+            2,
+            {
+                "x": {
```
Comment (dustjr1565) [inline tests/test_hpo.py:244]: key 에 있는 알파벳들은 뭘 의미하는건가요?

File: waffle_hub/core/hpo/adapter/optuna/optuna_hpo.py [hunk 2]
```diff
@@ -0,0 +1,431 @@
...
+    def metric(self) -> str:
+        """HPO metric"""
+        return self.__metric
+
+    @metric.setter
+    @type_validator(str)
+    def metric(self, v):
+        if v is None:
+            warnings.warn("HPO metric is not set. Set to None.")
+        self.__metric = v
```
Comment (dustjr1565) [inline waffle_hub/core/hpo/adapter/optuna/optuna_hpo.py:181]: If it's 'None', what criteria do you use for HPO?

File: waffle_hub/hub/hub.py [hunk 3]
```diff
@@ -1768,3 +1775,118 @@ def export_waffle(self, root_dir: str = None) -> ExportWaffleResult:
...
+        Args:
+            dataset (Union[Dataset, str], optional): The dataset to use for HPO. Can be a `Dataset` object or the name of a dataset. Defaults to None.
+            dataset_root_dir (Union[str, Path], optional): The root directory of the dataset. Defaults to None.
+            sampler (Union[dict, str], optional): The sampler to use for HPO. Can be a dictionary of sampler parameters or the name of a built-in sampler. Defaults to None.
+            pruner (Union[dict, str], optional): The pruner to use for HPO. Can be a dictionary of pruner parameters or the name of a built-in pruner. Defaults to None.
+            direction (str, optional): The direction of optimization. Can be 'maximize' or 'minimize'. Defaults to None.
+            n_trials (int, optional): The number of trials to run for HPO. Defaults to None.
+            metric (str, optional): The metric to optimize for. Defaults to None.
+            search_space (dict, optional): The search space for HPO. Defaults to None.
+            **kwargs: Additional keyword arguments to pass to the HPO function.
```
Comment (Jiyong-Jeon) [inline waffle_hub/hub/hub.py:1838]: What is the default value?

An error occurs when I insert the None value for each parameter.
I think you should keep `Optional` and set the Default value.

File: tests/test_hpo.py (hunk 1 above)
Comment (kimdanni) [reply tests/test_hpo.py:244]: The key corresponds to the name of the parameter given in the search space. For example, in a simple function used as the optimization target function with a variable named 'x', the key passed would be 'x'.

File: waffle_hub/core/hpo/adapter/optuna/optuna_hpo.py (hunk 2 above)
Comment (kimdanni) [reply waffle_hub/core/hpo/adapter/optuna/optuna_hpo.py:181]: Referring to optuna_hpo.py, if the metric is not input when using the hub, an error is emitted, and when using an objective func other than the hub, None is given.
search space is a value that must be entered.
All other values ​​are saved in the default config saved in optuna config.

File: tests/test_cli.py [hunk 4]
```diff
@@ -250,6 +249,46 @@ def test_hub_train_advance_params(test_dir: Path):
...
+        --name test_hpo \
+        --dataset {test_dir / 'datasets' / 'from_coco'} \
+        --sampler RandomSampler \
+        --pruner MedianPruner \
+        --direction maximize \
+        --n-trials 2 \
+        --metric accuracy \
+        --search_space {test_dir / 'search_space.json'} \
+    "
+
```
[the remark this answers is not in the document]
Comment (kimdanni) [reply tests/test_cli.py:282]: Please refer to the cli test code, hub_test. I have created a scenario where optimal parameters are obtained and trained by performing hpo before train.
If you look at train in hub.py, when an hpo file exists, the trial with the highest performance is stored in the hub that performed hpo, and this is imported into train config.

File: tests/test_hpo.py [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+        image_size=64,
+    )
+
+    hpo_train_config = TrainConfig.load(Path(hub.root_dir / hub.name / "configs" / "train.yaml"))
+    train_hub = Hub.load(name=name, root_dir=tmpdir)
+    train_hub: Hub = Hub.from_model_config(
+        name=name + "_from_model_config",
+        model_config_file=tmpdir / name / Hub.MODEL_CONFIG_FILE,
+        root_dir=tmpdir,
+    )
```
[the remark this answers is not in the document]
Comment (kimdanni) [reply tests/test_hpo.py:130]: modified

File: tests/test_hpo.py [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+        name=name,
+        backend="ultralytics",
+        task=TaskType.OBJECT_DETECTION,
+        model_type="yolov8",
+        model_size="n",
+        categories=dataset.get_category_names(),
+        root_dir=tmpdir,
+        device="cpu",
+        workers=0,
+        hold=True,
```
[the remark this answers is not in the document]
Comment (kimdanni) [reply tests/test_hpo.py:110]: modified

File: tests/test_hpo.py [hunk 7]
```diff
@@ -0,0 +1,339 @@
...
+        dataset=dataset,
+        epochs=hpo_train_config.epochs,
+        batch_size=hpo_train_config.batch_size,
+        image_size=hpo_train_config.image_size,
+        learning_rate=hpo_train_config.learning_rate,
+        letter_box=hpo_train_config.letter_box,
+        device="cpu",
+        workers=0,
+        advance_params=hpo_train_config.advance_params,
+    )
```
[the remark this answers is not in the document]
Comment (kimdanni) [reply tests/test_hpo.py:141]: I added code to the train function within the hub and modified it to enable learning with optimal hyperparameters automatically after hpo.

File: waffle_hub/hub/hub.py [hunk 8]
```diff
@@ -1768,3 +1775,118 @@ def export_waffle(self, root_dir: str = None) -> ExportWaffleResult:
...
+        dataset: Union[Dataset, str] = None,
+        dataset_root_dir: Union[str, Path] = None,
+        sampler: Union[dict, str] = None,
+        pruner: Union[dict, str] = None,
+        direction: str = None,
+        n_trials: int = None,
+        metric: str = None,
+        search_space: dict = None,
+        **kwargs,
+    ) -> dict:
```
[the remark this answers is not in the document]
Comment (kimdanni) [reply waffle_hub/hub/hub.py:1825]: modified

File: waffle_hub/hub/hub.py (hunk 3 above)
Comment (kimdanni) [reply waffle_hub/hub/hub.py:1838]: Added default hpo config and modified to assign default value through setter.
